src/System.IO.Pipelines/System/IO/Pipelines/BufferSegment.cs

```diff
@@ -7,21 +7,40 @@
 
 namespace System.IO.Pipelines
 {
-    internal class BufferSegment
+    public interface IMemoryList<T>
+    {
+        Memory<T> Memory { get; }
+
+        IMemoryList<T> Rest { get; }
```

Comment:
I would also like to rename `Rest` to `Next` it's more in the spirit of linked lists.